design-documents/atomics-ABI.rst

```diff
@@ -0,0 +1,308 @@
+and is not subject to an ABI, it may still depend on other libraries or components that do have an ABI.
+
+
+Constrain optimisations
+~~~~~~~~~~~~~~~~~~~~~~~
+
+There have been several instances where optimisations have been incorrectly applied,
+or attempts to apply optimisations to atomic code generation that induce unexpected
+concurrent program behaviour. This has happened frequently enough that we need to
+collect these cases together to outline why they should not occur. For example
+
+Consider the following Concurrent Program::
+
+  // Shared-Memory Locations
+  _Atomic int* x;
+  _Atomic int* y;
+
+  // Memory Order Parameter
+  #define relaxed memory_order_relaxed
+  #define release memory_order_release
+  #define acquire memory_order_acquire
+
+  // Threads of Execution
+  void thread_0 () {
+    atomic_store_explicit(x,1,relaxed);
+    atomic_thread_fence(release);
+    atomic_store_explicit(y,1,relaxed);
+  }
+
+  void thread_1 () {
+    atomic_exchange_explicit(y,2,release);
+    atomic_thread_fence(acquire);
+    int r0 = atomic_load_explicit(x,relaxed);
+  }
+
+
+Under ISO C, the above Concurrent Program finishes execution in one of three
+possible outcomes (a reference for this notation is found here [PAPER_])::
+
+  { thread_1:r0=0; y=1; }
+  { thread_1:r0=1; y=1; }
+  { thread_1:r0=1; y=2; }
+
+In this case the value read by the exchange on ``thread_1`` is not used, and a
+compiler is free to remove references to unused data. It is not legal according
+to this ABI for a compliant implementation piler to translate the program into
+the following Assembly Sequences::
+
+  thread_0:
+    MOV W9,#1
+    STR W9,[X2]
+    DMB ISH
+    STR W3,[X4]
+
+  thread_1:
+    MOV W9,#2
+    SWP W9, WZR, [X2]
+    DMB ISHLD
+    LDR W3,[X4]
+
+where ``thread_0:X2`` contains the address of ``x``, ``thread_0:X4`` contains
+the address of ``y``, and ``thread_1:X2`` contains the address of ``y``,
+``thread_1:X4`` contains the address of ``x``.
+
+The ``exchange`` Atomic Operation is compiled to a ``SWP`` Assembly
+Instruction, where its destination register is the zero register ``WZR``. The 
+``acquire`` fence on ``thread_1`` is compiled to the ``DMB ISHLD`` Assembly 
+Instruction.
+
+Executing the compiled program on an Arm-based machine from a fixed initial
+state (where ``x`` and ``y`` are ``0``) produces one of the following outcomes,
+according to the AArch64 Memory Model contained in §B2 of the Arm Architecture
+Reference Manual [ARMARM_]::
+
+  { thread_1:r0=0; [y]=1; }
+  { thread_1:r0=0; [y]=2; } <-- Forbidden by source model, a bug!
+  { thread_1:r0=1; [y]=1; }
+  { thread_1:r0=1; [y]=2; }
+
+By comparing ``W3`` and the local variable ``r0`` of the original Concurrent
+Program we see there is one additional outcome of executing the compiled
+program that is not an outcome of executing the Concurrent Program. This is due
+to the fact that according to the Arm Architecture Reference Manual [ARMARM_] 
```

Comment:
the embedded quite is grammatically still part of the sentence. And so the sentence still needs to be grammatically correct. It isn't in this case.